quinn/src/endpoint.rs

```diff
@@ -449,14 +453,8 @@ impl State {
                                 buf,
                                 &mut response_buffer,
                             ) {
-                                Some(DatagramEvent::NewConnection(handle, conn)) => {
-                                    let conn = self.connections.insert(
-                                        handle,
-                                        conn,
-                                        self.socket.clone(),
-                                        self.runtime.clone(),
-                                    );
-                                    self.incoming.push_back(conn);
+                                Some(DatagramEvent::NewConnection(incoming)) => {
+                                    self.incoming.push_back((incoming, response_buffer));
```

Comment:
Do we have any limits on the length of `self.incoming`? This might be a memory exhaustion hazard, even though the individual elements shouldn't be large. Previously `ServerConfig::concurrent_connections` limited it, but now we don't check that until later.

`ServerConfig::concurrent_connections` would still be a natural conservative limit, though we'd need to expose a getter; perhaps a hidden one, similar to `get_max_udp_payload_size`. We could even make the limit exact by exposing the current active connection count (e.g. publicizing `quinn_proto::Endpoint::known_connections`, maybe renamed to `connection_count`), which could make `ConnectionLimitExceeded` errors from `accept` unreachable via the `quinn` API, though I think the benefit is small.

Alternatively, we could just set a generous-but-finite ad-hoc threshold; say, 2^16 elements? That should be low enough to prevent memory from growing beyond 100MiB, and high enough not to lose connections except under load conditions that are catastrophic regardless. This would be simplest and I don't think it has much of a drawback.

Reply:
Hm... Well, if you say that a 2^16 limit seems good enough for now, then it is definitely the simplest, so I think I'll go with that. Nevertheless, there is an alternative approach (I'm a bit unclear on whether this is what you're suggesting in the second paragraph) where we count pending incoming connections towards the connection limit on the quinn-proto level. This diff would illustrate basically how we could do that:

<details><summary>Large diff</summary>

```diff
diff --git a/quinn-proto/src/endpoint.rs b/quinn-proto/src/endpoint.rs
index 7ce4d944..89a5c8b0 100644
--- a/quinn-proto/src/endpoint.rs
+++ b/quinn-proto/src/endpoint.rs
@@ -40,6 +40,7 @@ pub struct Endpoint {
     rng: StdRng,
     index: ConnectionIndex,
     connections: Slab<ConnectionMeta>,
+    num_incoming_connections: usize,
     local_cid_generator: Box<dyn ConnectionIdGenerator>,
     config: Arc<EndpointConfig>,
     server_config: Option<Arc<ServerConfig>>,
@@ -63,6 +64,7 @@ impl Endpoint {
             rng: rng_seed.map_or(StdRng::from_entropy(), StdRng::from_seed),
             index: ConnectionIndex::default(),
             connections: Slab::new(),
+            num_incoming_connections: 0,
             local_cid_generator: (config.connection_id_generator_factory.as_ref())(),
             config,
             server_config,
@@ -434,14 +436,23 @@ impl Endpoint {
             return None;
         }
 
-        if let Err(initial_close) =
-            self.check_connection_limit(version, addresses, &crypto, &src_cid, buf)
+        let server_config = self.server_config.as_ref().unwrap();
+
+        if (self.connections.len() + self.num_incoming_connections)
+            >= server_config.concurrent_connections as usize
+            || self.is_full()
         {
-            return Some(DatagramEvent::Response(initial_close));
+            debug!("refusing connection");
+            return Some(DatagramEvent::Response(self.initial_close(
+                version,
+                addresses,
+                &crypto,
+                &src_cid,
+                TransportError::CONNECTION_REFUSED(""),
+                buf,
+            )));
         }
 
-        let server_config = self.server_config.as_ref().unwrap();
-
         if dst_cid.len() < 8 && dst_cid.len() != self.local_cid_generator.cid_len() {
             debug!(
                 "rejecting connection due to invalid DCID length {}",
@@ -485,6 +496,8 @@ impl Endpoint {
             }
         };
 
+        self.num_incoming_connections += 1;
+
         Some(DatagramEvent::NewConnection(IncomingConnection {
             addresses,
             ecn,
@@ -507,15 +520,7 @@ impl Endpoint {
         now: Instant,
         buf: &mut BytesMut,
     ) -> Result<(ConnectionHandle, Connection), (ConnectionError, Option<Transmit>)> {
-        self.check_connection_limit(
-            incoming.version,
-            incoming.addresses,
-            &incoming.crypto,
-            &incoming.src_cid,
-            buf,
-        )
-        .map_err(|response| (ConnectionError::ConnectionLimitExceeded, Some(response)))?;
-
+        self.num_incoming_connections -= 1;
         let server_config = self.server_config.as_ref().unwrap().clone();
 
         let ch = ConnectionHandle(self.connections.vacant_key());
@@ -585,6 +590,7 @@ impl Endpoint {
 
     /// Reject this incoming connection attempt
     pub fn reject(&mut self, incoming: IncomingConnection, buf: &mut BytesMut) -> Transmit {
+        self.num_incoming_connections -= 1;
         self.initial_close(
             incoming.version,
             incoming.addresses,
@@ -606,6 +612,7 @@ impl Endpoint {
         if incoming.remote_address_validated() {
             return Err(RetryError(incoming));
         }
+        self.num_incoming_connections -= 1;
         let server_config = self.server_config.as_ref().unwrap();
 
         // First Initial
@@ -653,6 +660,11 @@ impl Endpoint {
         })
     }
 
+    /// Call when ignoring an incoming connection attempt to stop counting it towards the connection limit
+    pub fn ignore_incoming(&mut self) {
+        self.num_incoming_connections -= 1;
+    }
+
     fn add_connection(
         &mut self,
         ch: ConnectionHandle,
@@ -701,30 +713,6 @@ impl Endpoint {
         conn
     }
 
-    fn check_connection_limit(
-        &mut self,
-        version: u32,
-        addresses: FourTuple,
-        crypto: &Keys,
-        src_cid: &ConnectionId,
-        buf: &mut BytesMut,
-    ) -> Result<(), Transmit> {
-        let server_config = self.server_config.as_ref().unwrap();
-        if self.connections.len() < server_config.concurrent_connections as usize && !self.is_full()
-        {
-            return Ok(());
-        }
-        debug!("refusing connection");
-        Err(self.initial_close(
-            version,
-            addresses,
-            crypto,
-            src_cid,
-            TransportError::CONNECTION_REFUSED(""),
-            buf,
-        ))
-    }
-
     fn initial_close(
         &mut self,
         version: u32,
@@ -957,6 +945,11 @@ pub enum DatagramEvent {
     /// The datagram is redirected to its `Connection`
     ConnectionEvent(ConnectionHandle, ConnectionEvent),
     /// The datagram may result in starting a new `Connection`
+    ///
+    /// This event corresponds to an incrementing of the incoming connection counter, which counts
+    /// towards the active connection limit. Therefore, this incoming connection should not be
+    /// dropped without a corresponding call to `accept`, `reject`, `retry`, or `ignore_incoming`,
+    /// so as to decrement the incoming connection counter.
     NewConnection(IncomingConnection),
     /// Response generated directly by the endpoint
     Response(Transmit),
diff --git a/quinn/src/endpoint.rs b/quinn/src/endpoint.rs
index d962c7d4..119504e2 100644
--- a/quinn/src/endpoint.rs
+++ b/quinn/src/endpoint.rs
@@ -815,4 +815,8 @@ impl EndpointInner {
             Err(e) => Err((e, response_buffer)),
         }
     }
+
+    pub(crate) fn ignore_incoming(&self) {
+        self.state.lock().unwrap().inner.ignore_incoming();
+    }
 }
diff --git a/quinn/src/incoming_connection.rs b/quinn/src/incoming_connection.rs
index 247cf751..9376b426 100644
--- a/quinn/src/incoming_connection.rs
+++ b/quinn/src/incoming_connection.rs
@@ -87,7 +87,7 @@ impl IncomingConnection {
 
     /// Ignore this incoming connection attempt, not sending any packet in response
     pub fn ignore(mut self) {
-        self.0.take().unwrap();
+        self.0.take().unwrap().endpoint.ignore_incoming();
     }
 }
 
```

</details>

But it does introduce a very easy to make and annoying to catch possible usage error of quinn-proto wherein the user discards an `quinn_proto::IncomingConnection` in one case or another without somehow triggering a decrement `num_incoming_connections` (or otherwise fails to keep that counter synchronized with reality), and there are other places in quinn where these sorts of ad-hoc heuristic numbers seem to exist, so I think for now I'll just do your final suggestion yeah, but this is here as a note if we feel a need to re-visit this in the future I guess.

Reply:
Done.